src/include/interval_set.h

```diff
@@ -123,84 +127,83 @@ class interval_set {
         }
 
         // Dereference this iterator to get a pair.
-        std::pair < T, T > operator*() const {
-                return *_iter;
+        map_const_reference& operator*() const {
+          return *_iter;
         }
 
         // Return the interval start.
-        T get_start() const {
-                return _iter->first;
+        value_type get_start() const {
+          return _iter->first;
         }
-        T get_end() const {
-                return _iter->first + _iter->second;
+        value_type get_end() const {
+          return _iter->first + _iter->second;
         }
 
         // Return the interval length.
-        T get_len() const {
-                return _iter->second;
+        value_type get_len() const {
+          return _iter->second;
         }
 
         // Preincrement
-        const_iterator &operator++()
+        const_iterator& operator++()
         {
-                ++_iter;
-                return *this;
+          ++_iter;
+          return *this;
         }
 
         // Postincrement
         const_iterator operator++(int)
         {
-                const_iterator prev(_iter);
-                ++_iter;
-                return prev;
+          const_iterator prev(_iter);
+          ++_iter;
+          return prev;
         }
 
     protected:
         typename Map::const_iterator _iter;
   };
 
-  interval_set() : _size(0) {}
+  interval_set() = default;
   interval_set(Map&& other) {
     m.swap(other);
-    _size = 0;
-    for (auto& i : m) {
+    for (const auto& i : m) {
       _size += i.second;
     }
   }
 
-  int num_intervals() const
+  size_type num_intervals() const
   {
     return m.size();
   }
 
-  typename interval_set<T,Map>::iterator begin() {
+  auto begin() {
     return typename interval_set<T,Map>::iterator(m.begin());
   }
 
-  typename interval_set<T,Map>::iterator lower_bound(T start) {
+  auto lower_bound(T start) {
     return typename interval_set<T,Map>::iterator(find_inc_m(start));
   }
 
-  typename interval_set<T,Map>::iterator end() {
+  auto end() {
     return typename interval_set<T,Map>::iterator(m.end());
   }
 
-  typename interval_set<T,Map>::const_iterator begin() const {
+  auto begin() const {
```

Comment:
i'd suggest s/auto/const_iterator/. IMHO, this hurts the readability of this source file.